pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
+	if queue.Len() < e.activeWorkers {
+		time.Sleep(requeuePeriod)
 	}
-	secrets, err := e.listTokenSecrets(serviceAccount)
-	if err != nil {
-		glog.Error(err)
-		return
+	queue.Add(key)
+}
+
+func (e *TokensController) queueSecretSync(obj interface{}) {
+	if secret, ok := obj.(*api.Secret); ok {
+		e.syncSecretQueue.Add(makeSecretKey(secret))
 	}
-	for _, secret := range secrets {
-		glog.V(4).Infof("Deleting secret %s/%s because service account %s was deleted", secret.Namespace, secret.Name, serviceAccount.Name)
-		if err := e.deleteSecret(secret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", secret.Namespace, secret.Name, err)
-		}
+}
+
+func (e *TokensController) queueSecretUpdateSync(oldObj interface{}, newObj interface{}) {
+	if secret, ok := newObj.(*api.Secret); ok {
+		e.syncSecretQueue.Add(makeSecretKey(secret))
 	}
 }
 
-// secretAdded reacts to a Secret create by ensuring the referenced ServiceAccount exists, and by adding a token to the secret if needed
-func (e *TokensController) secretAdded(obj interface{}) {
-	secret := obj.(*api.Secret)
-	serviceAccount, err := e.getServiceAccount(secret, true)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) syncServiceAccount() {
+	key, quit := e.syncServiceAccountQueue.Get()
+	if quit {
 		return
 	}
-	if serviceAccount == nil {
-		glog.V(2).Infof(
-			"Deleting new secret %s/%s because service account %s (uid=%s) was not found",
-			secret.Namespace, secret.Name,
-			secret.Annotations[api.ServiceAccountNameKey], secret.Annotations[api.ServiceAccountUIDKey])
-		if err := e.deleteSecret(secret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", secret.Namespace, secret.Name, err)
-		}
-	} else {
-		e.generateTokenIfNeeded(serviceAccount, secret)
-	}
-}
+	defer e.syncServiceAccountQueue.Done(key)
 
-// secretUpdated reacts to a Secret update (or re-list) by deleting the secret (if the referenced ServiceAccount does not exist)
-func (e *TokensController) secretUpdated(oldObj interface{}, newObj interface{}) {
-	newSecret := newObj.(*api.Secret)
-	newServiceAccount, err := e.getServiceAccount(newSecret, true)
+	ns, name, uid, err := parseServiceAccountKey(key)
 	if err != nil {
 		glog.Error(err)
 		return
 	}
-	if newServiceAccount == nil {
-		glog.V(2).Infof(
-			"Deleting updated secret %s/%s because service account %s (uid=%s) was not found",
-			newSecret.Namespace, newSecret.Name,
-			newSecret.Annotations[api.ServiceAccountNameKey], newSecret.Annotations[api.ServiceAccountUIDKey])
-		if err := e.deleteSecret(newSecret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", newSecret.Namespace, newSecret.Name, err)
+
+	sa, err := e.getServiceAccount(ns, name, uid, false)
+	switch {
+	case err != nil:
+		// TODO: requeue?
+		glog.Error(err)
+	case sa == nil:
+		// service account no longer exists, so delete related tokens
+		glog.V(4).Infof("syncServiceAccount(%s/%s), service account deleted, removing tokens", ns, name)
```

Comment:
How is this behavior triggered on controller startup (i.e., if you crash after the account is deleted but before this code manages to run?)